strutil/quantity/quantity.go

```diff
@@ -200,3 +201,247 @@ func FormatDuration(dt float64) string {
+// rendering, we either need to peform a ceiling, floor or rounding
+// operation with the remainder below the minimum.
+type RenderMode uint32
+
+const (
+	TimeLeft    RenderMode = 0
+	TimePassed  RenderMode = 1
+	TimeRounded RenderMode = 2
+)
+
+// FormatDurationGeneric formats time duration similar to systemd
+// format_timespan, but with options allowing for a more compact
+// arrangement, including shortened suffixes. The function is provided
+// with dt in seconds (as in the output of time.Now().Seconds()). The 'min'
+// and 'max' Duration allows limiting the unit range rendered. Place values
+// less than 'min' will be processed according to RenderMode. RenderMode
+// allows rounding, flooring or ceiling to be applied to place value below
+// 'min'. Durations above the 'max' place value will result in "ages!".
+func FormatDurationGeneric(dt float64, min Duration, max Duration, count Places, mode RenderMode, space SpaceMode) string {
+	var units = map[Duration]string{
+		Year: "y",
+		Day:  "d",
+		Hour: "h",
+		Min:  "m",
+		Sec:  "s",
+		MSec: "ms",
+		USec: "µs",
+		NSec: "ns",
+	}
+
+	// We need to access the map in order using an ordered array
+	var keys []Duration
+	for k := range units {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
+
+	var render string
+
+	// Invalid case: min > max
+	if min > max {
+		return "inv!"
+	}
+
+	// Special case: zero duration
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := uint64(dt * float64(Sec))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowComp {
+		for i := 0; i < len(keys); i++ {
+			// Only render places as indicated by count
+			if delta >= uint64(keys[i]) && keys[i] >= Min {
+				index := i + int(count) - 1
+				if keys[index] > min {
+					min = keys[index]
+				}
+				break
+			}
+
+			// Only one place as we are s, ms, us or ns
+			if delta >= uint64(keys[i]) && keys[i] < Min {
+				if keys[i] > min {
+					min = keys[i]
+				}
+				break
+			}
+		}
+	}
+
+	// Apply the rendermode to the least significant place value
+	mod := uint64(delta % uint64(min))
+	switch {
+	case mode == TimeLeft && mod > 0:
+		// Ceiling
+		delta = delta + uint64(min) - mod
+	case mode == TimePassed:
+		// Floor
+		delta = delta - mod
+	case mode == TimeRounded && mod >= (uint64(min)/2):
+		// Rounded Up
+		delta = delta + uint64(min) - mod
+	case mode == TimeRounded && mod < (uint64(min)/2):
+		// Rounded Down
+		delta = delta - mod
+	}
+
+	// Special case: less than minimum accuracy
+	if delta < uint64(min) {
+		return "0" + units[min]
+	}
+
+	// In ShowComp mode only 2-digit days are possible
+	if count == ShowComp && delta >= (100*uint64(Day)) {
```

Comment:
it's not clear to me, how do we know that we are dealing with days (and not years, for example)?

Reply:
The casting is now removed, as per your previous suggestion (I think that hide the "Day" duration part ;)

We are saying that if we are > 99 days, we will be needing 3 digits, which will not work for ShowCompact mode, so lets only support ShowCompact up until 99 days of total duration. ShowCompact mode is for progress indicators, and 99 days is plenty of time to wait staring at the progress bar.